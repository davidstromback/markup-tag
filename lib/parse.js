import { bodyOrTagOpener } from "./state.js";
export function parse(tokens) {
    const root = { type: "", props: {}, children: [] };
    const context = {
        attribute: undefined,
        closeMarker: false,
        stack: [root],
        state: bodyOrTagOpener,
        tag: { type: "", props: {}, children: [] },
        token: { type: "text", value: undefined },
    };
    for (const token of tokens) {
        context.token = token;
        context.state = context.state(context);
    }
    const element = context.stack.pop();
    if (context.stack.length !== 0 || element !== root) {
        throw new Error(`Expected closing tag for ${element === null || element === void 0 ? void 0 : element.type}.`);
    }
    return element.children;
}
