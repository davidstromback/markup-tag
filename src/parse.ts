import { bodyOrTagOpener } from "./state.js";
import { Token, Interpolation, ParseContext } from "./types.js";

export function parse(tokens: Iterable<Token | Interpolation>) {
  const root = { type: "", props: {}, children: [] };

  const context: ParseContext = {
    attribute: undefined,
    closeMarker: false,
    stack: [root],
    state: bodyOrTagOpener,
    tag: { type: "", props: {}, children: [] },
    token: { type: "text", value: undefined },
  };

  for (const token of tokens) {
    context.token = token;
    context.state = context.state(context);
  }

  const element = context.stack.pop();

  if (context.stack.length !== 0 || element !== root) {
    throw new Error(`Expected closing tag for ${element?.type}.`);
  }

  return element.children;
}
