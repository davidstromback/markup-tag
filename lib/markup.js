import { parse } from "./parse.js";
import { interpolate } from "./interpolate.js";
import { compile, tokenizer, memo } from "fsm-tokenizer";
export const schema = compile({
    initialState: "text",
    tokens: {
        closeMarker: {},
        equals: {},
        escape: {},
        name: { value: true },
        quote: {},
        tagCloser: {},
        tagOpener: {},
        text: { value: true },
        value: { value: true, padding: true },
    },
    states: {
        text: {
            token: "text",
            rules: [
                [/\\/, "before", "textEscape", true],
                [/</, "before", "tagOpener", true],
                [/[\s\S]/, "after", "text", false],
            ],
        },
        textEscape: {
            token: "escape",
            rules: [[/\\/, "after", "textEscaped", true]],
        },
        textEscaped: {
            token: "text",
            rules: [[/[\s\S]/, "after", "text", false]],
        },
        tagOpener: {
            token: "tagOpener",
            rules: [[/</, "after", "nameOrClose", true]],
        },
        closeMarker: {
            token: "closeMarker",
            rules: [[/\//, "after", "nameOrClose", true]],
        },
        nameOrClose: {
            token: "name",
            rules: [
                [/\s/, "after", "nameOrClose", false],
                [/[a-zA-Z]/, "before", "name", true],
                [/\//, "before", "closeMarker", true],
                [/>/, "before", "tagCloser", true],
            ],
        },
        name: {
            token: "name",
            rules: [
                [/[a-zA-Z0-9]/, "after", "name", false],
                [/\s/, "before", "nameOrClose", true],
                [/>/, "before", "tagCloser", true],
                [/\//, "before", "closeMarker", true],
                [/=/, "before", "equalsOrName", true],
            ],
        },
        equalsOrName: {
            token: "equals",
            rules: [
                [/\s/, "after", "equalsOrName", false],
                [/=/, "after", "openingQuoteOrName", true],
                [/\//, "before", "closeMarker", true],
                [/>/, "before", "tagCloser", true],
                [/[a-zA-Z]/, "before", "nameOrClose", true],
            ],
        },
        openingQuoteOrName: {
            token: "quote",
            rules: [
                [/\s/, "after", "openingQuoteOrName", false],
                [/"/, "after", "value", true],
                [/\//, "before", "closeMarker", true],
                [/>/, "before", "tagCloser", true],
                [/[a-zA-Z]/, "before", "nameOrClose", true],
            ],
        },
        value: {
            token: "value",
            rules: [
                [/\\/, "before", "valueEscape", true],
                [/"/, "before", "closingQuote", true],
                [/[\s\S]/, "after", "value", false],
            ],
        },
        valueEscape: {
            token: "escape",
            rules: [[/\\/, "after", "valueEscaped", true]],
        },
        valueEscaped: {
            token: "value",
            rules: [[/[\s\S]/, "after", "value", false]],
        },
        closingQuote: {
            token: "quote",
            rules: [[/"/, "after", "nameOrClose", true]],
        },
        tagCloser: {
            token: "tagCloser",
            rules: [[/>/, "after", "text", true]],
        },
    },
});
const tokenize = memo(tokenizer(schema, (type, value) => ({ type, value })));
export function markup(strings, ...values) {
    return parse(interpolate(tokenize, strings, values));
}
