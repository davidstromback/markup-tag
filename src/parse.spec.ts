import { tokenizer } from "fsm-tokenizer";
import { schema } from "./markup.js";
import { parse } from "./parse.js";

describe("parse", () => {
  describe("matches snapshots", () => {
    const cases = [
      "<foobar />",
      '<foo bar="foo" />',
      "<foo>Hello world</foo>",
      "<foo><foobar /></foo>",
    ];

    test.each(cases)("%p", (string) => {
      const tokenize = tokenizer(schema);
      expect(parse(tokenize(string))).toMatchSnapshot();
    });
  });

  /*
  describe("throws errors matching snapshots", () => {
    const cases: string[] = [];

    test.each(cases)("%p", (string) => {
      expect(parse(string)).toThrowErrorMatchingSnapshot();
    });
  });
  */
});
