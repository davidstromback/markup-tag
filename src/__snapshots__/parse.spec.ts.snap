// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse matches snapshots "<foo bar=\\"foo\\" />" 1`] = `
Array [
  Object {
    "children": Array [],
    "props": Object {
      "bar": "foo",
    },
    "type": "foo",
  },
]
`;

exports[`parse matches snapshots "<foo><foobar /></foo>" 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "props": Object {},
        "type": "foobar",
      },
    ],
    "props": Object {},
    "type": "foo",
  },
]
`;

exports[`parse matches snapshots "<foo>Hello world</foo>" 1`] = `
Array [
  Object {
    "children": Array [
      "Hello world",
    ],
    "props": Object {},
    "type": "foo",
  },
]
`;

exports[`parse matches snapshots "<foobar />" 1`] = `
Array [
  Object {
    "children": Array [],
    "props": Object {},
    "type": "foobar",
  },
]
`;
